File: common/src/main/java/cc/turtl/chiselmon/feature/pc/tooltip/TooltipRegistry.java
package cc.turtl.chiselmon.feature.pc.tooltip;

import cc.turtl.chiselmon.client.config.category.PCConfig;
import cc.turtl.chiselmon.core.api.predicate.PokemonPredicatesKt;
import cc.turtl.chiselmon.util.format.PokemonFormats;
import com.cobblemon.mod.common.pokemon.Pokemon;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;

public final class TooltipRegistry {
    private static final List<TooltipEntry> ENTRIES = new ArrayList<>();

    static {
        // Register standard entries
        add("ivs", PCConfig.TooltipConfig::getIvs, p -> true, PokemonFormats::ivsSummary);
        add("original_trainer", PCConfig.TooltipConfig::getOriginalTrainer, p -> true, Pokemon::getOriginalTrainerName);
        add("form", PCConfig.TooltipConfig::getForm, p -> true, p -> p.getForm().getName());
        add("friendship", PCConfig.TooltipConfig::getFriendship, p -> true, Pokemon::getFriendship);

        // Register conditional entries
        add("ride_styles", PCConfig.TooltipConfig::getRideStyles, PokemonPredicatesKt.IS_RIDEABLE, PokemonFormats::rideStyles);
        add("marks", PCConfig.TooltipConfig::getMarks, PokemonPredicatesKt.IS_MARKED, PokemonFormats::marks);
        add("hatch_progress", PCConfig.TooltipConfig::getHatchProgress, PokemonPredicatesKt.IS_EGG_DUMMY, PokemonFormats::hatchProgress);
    }

    private TooltipRegistry() {
    }

    private static void add(String key, Predicate<PCConfig.TooltipConfig> configCheck, Predicate<Pokemon> pokemonCheck, Function<Pokemon, Object> val) {
        ENTRIES.add(new TooltipEntry(key, configCheck, pokemonCheck, val));
    }

    public static List<TooltipEntry> getEntries() {
        return ENTRIES;
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/feature/pc/tooltip/TooltipRegistry.kt
package cc.turtl.chiselmon.client.feature.pc.tooltip

import cc.turtl.chiselmon.client.config.category.PCConfig
import cc.turtl.chiselmon.core.api.predicate.IS_EGG_DUMMY
import cc.turtl.chiselmon.core.api.predicate.IS_MARKED
import cc.turtl.chiselmon.core.api.predicate.IS_RIDEABLE
import cc.turtl.chiselmon.core.util.format.PokemonFormats
import com.cobblemon.mod.common.pokemon.Pokemon

object TooltipRegistry {
    private val entries = listOf(
        entry("ivs", { it.ivs }, { true }, PokemonFormats::ivsSummary),
        entry("original_trainer", { it.originalTrainer }, { true }, { it.originalTrainerName ?: "???" }),
        entry("form", { it.form }, { true }, { it.form.name }),
        entry("friendship", { it.friendship }, { true }, { it.friendship }),
        entry("ride_styles", { it.rideStyles }, IS_RIDEABLE::test, PokemonFormats::rideStyles),
        entry("marks", { it.marks }, IS_MARKED::test, PokemonFormats::marks),
        entry("hatch_progress", { it.hatchProgress }, IS_EGG_DUMMY::test, PokemonFormats::hatchProgress),
    )

    @JvmStatic
    fun getEntries(): List<TooltipEntry> = entries

    private fun entry(
        key: String,
        configCheck: (PCConfig.TooltipConfig) -> Boolean,
        pokemonCheck: (Pokemon) -> Boolean,
        provider: (Pokemon) -> Any
    ) = TooltipEntry(key, configCheck, pokemonCheck, provider)
}